subsys/bluetooth/services/fast_pair/fp_keys.c

```diff
@@ -295,11 +347,56 @@ int fp_keys_store_account_key(const struct bt_conn *conn, const struct fp_accoun
 	return err;
 }
 
+int fp_keys_store_personalized_name(const struct bt_conn *conn, const char *pn)
+{
+	struct fp_procedure *proc = &fp_procedures[bt_conn_index(conn)];
+	int err = 0;
+
+	if (!(proc->wait_for_mask & BIT(WAIT_FOR_PERSONALIZED_NAME_BIT_POS))) {
+		LOG_WRN("Invalid state during Personalized Name write");
+		return -EACCES;
+	}
+	WRITE_BIT(proc->wait_for_mask, WAIT_FOR_PERSONALIZED_NAME_BIT_POS, 0);
+
+	/* Invalidate Key to ensure that requests will be rejected until procedure is restarted. */
+	invalidate_key(proc);
+
+	err = fp_storage_pn_save(pn);
+	if (!err) {
+		LOG_DBG("Personalized Name stored");
+	} else {
+		LOG_ERR("Store Personalized Name error: err=%d", err);
+	}
+
+	return err;
+}
+
+int fp_keys_wait_for_personalized_name(const struct bt_conn *conn)
+{
+	struct fp_procedure *proc = &fp_procedures[bt_conn_index(conn)];
+
+	if (proc->state != FP_STATE_USE_TEMP_ACCOUNT_KEY) {
+		LOG_ERR("Invalid state to wait for personalized name");
+		return -EACCES;
+	}
+
+	__ASSERT_NO_MSG(!(proc->wait_for_mask & BIT(WAIT_FOR_ACCOUNT_KEY_BIT_POS)));
+
+	if (proc->wait_for_mask & BIT(WAIT_FOR_PERSONALIZED_NAME_BIT_POS)) {
+		LOG_ERR("Already waiting for personalized name");
+		return -EALREADY;
+	}
+
+	WRITE_BIT(proc->wait_for_mask, WAIT_FOR_PERSONALIZED_NAME_BIT_POS, 1);
+
+	return 0;
+}
+
 void fp_keys_bt_auth_progress(const struct bt_conn *conn, bool authenticated)
 {
 	struct fp_procedure *proc = &fp_procedures[bt_conn_index(conn)];
 
-	if (proc->state == FP_STATE_USE_TEMP_KEY) {
+	if (is_key_generated(proc->state)) {
```

Comment:
btw. This is expected only in the `FP_STATE_USE_TEMP_KEY` state. Could we bring back previous state check ?

Reply:
It's actually a point of concern. Should we allow writing account key after subsequent pairing (pairing when using account key in not discoverable advertising)? From logical point of view we shouldn't allow it but on the other hand, specification claims (not explicitly) that we should allow it.

Reply:
I double checked in the specification and it actually seems to allow for Account Key writes also on subsequent pairings. Let's leave it as it is now then.